Use NIO for file decryption

ehrmann opened the pull request:
Refactor the decryption code to use NIO classes. There's a pretty good performance win, and I think it's due to a combination of larger buffers and fewer intermediate buffers.

I tested it by extracting with the modified code and comparing the restored files.

To get an estimate of the performance win, I wrote quick test script that uses similar decrypt implementations over different file sizes:

https://gist.github.com/ehrmann/e80e4eda48b839d41d9cd52d0ba551ee

Windows:
```
 size,       old,       new
   1K,      1969,      1399
   2K,      2066,      1339
   4K,      1893,      1357
   8K,      1921,      1355
  16K,      2172,      1390
  32K,      2270,      1484
  64K,      2632,      1518
 128K,      2767,      1684
 256K,      3469,      1962
 512K,      5074,      2643
1024K,      7862,      3724
2048K,     13456,      6070
4096K,     26723,     12902
```

FreeBSD
```
 size,       old,       new
   1K,        60,        42
   2K,        96,        21
   4K,        17,        42
   8K,        34,        28
  16K,        44,        37
  32K,        83,        55
  64K,       120,        78
 128K,       189,       135
 256K,       397,       144
 512K,       306,       376
1024K,      1136,       591
2048K,      1444,       640
4096K,      3563,      2326
```

Linux
```
 size,       old,       new
   1K,       105,        35
   2K,       107,        34
   4K,       107,        32
   8K,       120,        42
  16K,       127,        40
  32K,       165,        47
  64K,       192,        64
 128K,       278,        98
 256K,       438,       160
 512K,       773,       292
1024K,      1422,       525
2048K,      2680,      1047
4096K,      5207,      2020
```

ehrmann committed:
Use NIO for file decryption

ehrmann committed:
Add test for KeyBag

MaxiHuHe04 commented:
Hi @ehrmann,
First of all, apologies for the long delay. Thank you for your contribution.
In my testing, the results vary heavily based on disks and caching. The larger buffer size is a good idea (the default for buffered streams is 8192 apparently). I got similar timings by just adding the same buffer size as second argument to BufferedInputStream and BufferedOutputStream however. So I don't think that NIO itself has a large performance benefit in this case, though I only tested that on Windows.
I also noticed a little bug in your CryptoBench. In decryptNew, you don't use dest but write to the input file. But that doesn't change the results much on my system.
I don't know how much this will change real-world performance because I am always limited by disk speed the first time I export something. Nevertheless, I think using NIO doesn't hurt

MaxiHuHe04 commented:
I noticed something else now which makes this a bit more complicated. I would like to fix the file paddings (#66), but I think this is much easier with cipher streams and just using transferTo. Currently, your code always writes one buffer size too much, I think because there is a inBuffer.flip() missing before the doFinal. This doesn't matter at the moment because it is immediately truncated again. But when using PKCS5Padding, it seems that we would also need to implement much of the functionality of CipherInputStream ourselves to ensure enough buffer capacity.

ehrmann committed:
Fix broken file padding logic

ehrmann commented:
Good catch on the missing `flip()`! I added it.

I read over #66 and added what I think is the `PKCS5Padding` version of the fix as a new commit. It just adds that padding to the cipher, but I think that should be enough.

MaxiHuHe04 commented:
Hi, does this work on your machine? I get a `javax.crypto.ShortBufferException: Need at least 16400 bytes of space in output buffer` when trying to decrypt the database of any of my backups.
This is also what I was worried about when I wrote this:
> But when using PKCS5Padding, it seems that we would also need to implement much of the functionality of CipherInputStream ourselves to ensure enough buffer capacity.

I think the problem is that you would have to dynamically adjust the buffer size depending on what number of padding bytes encrypted files have.
So for now, I will fix the padding corruption issue with PKCS5Padding, the current transferTo method, and larger buffer sizes. I don't think the manual use of ByteBuffers has a big advantage. Nevertheless, thank you for your work on this. If you find out something else, feel free to re-open the PR.
I would also be happy to just add your decryption test if you increased the size a bit to a few buffer sizes, maybe testing around various powers of 2. With only up to 128 bytes, it didn't fail with your current changes.

ehrmann commented:
> Hi, does this work on your machine?

Ah. I didn't; I thought the unit tests would catch it, but they didn't cover enough size ranges.

I updated the branch with a fix, ran it locally, and verified the output was the same.

> I don't think the manual use of ByteBuffers has a big advantage.

I just ran it on a backup and extracted ~60GB from the camera roll. A modern macbook showed a negligible difference, and an x86 Linux VM showed an **8% performance improvement** with the nio version. **This might not be interesting enough to bother with.** The win is from removing one or two buffer copies. Disk is slower than memory, so disk and decryption dominate wall time performance.
